<!DOCTYPE html>
<html lang="zh-cn" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">
<title>“抛出exception”与“传递参数”或“调用虚函数”之间的区别 | Forz Blog</title>
<meta name="keywords" content="" />
<meta name="description" content="从语法上看，在函数里声明参数与在 catch 子句中声明参数几乎没有什么差别: 1 2 3 4 5 6 7 8 9 10 11 class Widget { ... }; void f1(Widget w); void f2(Widget&amp; w); void f3(const Widget&amp; w); void f4(Widget *pw); void f5(const Widget *pw); catch (Widget w) ...">
<meta name="author" content="">
<link rel="canonical" href="/post/%E6%8A%9B%E5%87%BAexception%E4%B8%8E%E4%BC%A0%E9%80%92%E5%8F%82%E6%95%B0%E6%88%96%E8%B0%83%E7%94%A8%E8%99%9A%E5%87%BD%E6%95%B0%E4%B9%8B%E9%97%B4%E7%9A%84%E5%8C%BA%E5%88%AB/" />
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.00d5d4fc479b1575183ee8d86b4fb372ba9d9b1904e96fa8e4c40ff7debe2b94.css" integrity="sha256-ANXU/EebFXUYPujYa0&#43;zcrqdmxkE6W&#43;o5MQP996&#43;K5Q=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.7680afc38aa6b15ddf158a4f3780b7b1f7dde7e91d26f073e6229bb7a0793c92.js" integrity="sha256-doCvw4qmsV3fFYpPN4C3sffd5&#43;kdJvBz5iKbt6B5PJI="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="/favicon.ico">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="apple-touch-icon" href="/apple-touch-icon.png">
<link rel="mask-icon" href="/safari-pinned-tab.svg">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<meta name="generator" content="Hugo 0.87.0" />
<meta property="og:title" content="“抛出exception”与“传递参数”或“调用虚函数”之间的区别" />
<meta property="og:description" content="从语法上看，在函数里声明参数与在 catch 子句中声明参数几乎没有什么差别: 1 2 3 4 5 6 7 8 9 10 11 class Widget { ... }; void f1(Widget w); void f2(Widget&amp; w); void f3(const Widget&amp; w); void f4(Widget *pw); void f5(const Widget *pw); catch (Widget w) ..." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/post/%E6%8A%9B%E5%87%BAexception%E4%B8%8E%E4%BC%A0%E9%80%92%E5%8F%82%E6%95%B0%E6%88%96%E8%B0%83%E7%94%A8%E8%99%9A%E5%87%BD%E6%95%B0%E4%B9%8B%E9%97%B4%E7%9A%84%E5%8C%BA%E5%88%AB/" /><meta property="article:section" content="post" />
<meta property="article:published_time" content="2017-09-01T17:18:07&#43;00:00" />
<meta property="article:modified_time" content="2017-09-01T17:18:07&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="“抛出exception”与“传递参数”或“调用虚函数”之间的区别"/>
<meta name="twitter:description" content="从语法上看，在函数里声明参数与在 catch 子句中声明参数几乎没有什么差别: 1 2 3 4 5 6 7 8 9 10 11 class Widget { ... }; void f1(Widget w); void f2(Widget&amp; w); void f3(const Widget&amp; w); void f4(Widget *pw); void f5(const Widget *pw); catch (Widget w) ..."/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "Posts",
      "item": "/post/"
    }
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "“抛出exception”与“传递参数”或“调用虚函数”之间的区别",
      "item": "/post/%E6%8A%9B%E5%87%BAexception%E4%B8%8E%E4%BC%A0%E9%80%92%E5%8F%82%E6%95%B0%E6%88%96%E8%B0%83%E7%94%A8%E8%99%9A%E5%87%BD%E6%95%B0%E4%B9%8B%E9%97%B4%E7%9A%84%E5%8C%BA%E5%88%AB/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "“抛出exception”与“传递参数”或“调用虚函数”之间的区别",
  "name": "“抛出exception”与“传递参数”或“调用虚函数”之间的区别",
  "description": "从语法上看，在函数里声明参数与在 catch 子句中声明参数几乎没有什么差别: 1 2 3 4 5 6 7 8 9 10 11 class Widget { ... }; void f1(Widget w); void f2(Widget\u0026amp; w); void f3(const Widget\u0026amp; w); void f4(Widget *pw); void f5(const Widget *pw); catch (Widget w) ...",
  "keywords": [
    
  ],
  "articleBody": "从语法上看，在函数里声明参数与在 catch 子句中声明参数几乎没有什么差别:\n1 2 3 4 5 6 7 8 9 10 11  class Widget { ... }; void f1(Widget w); void f2(Widget\u0026 w); void f3(const Widget\u0026 w); void f4(Widget *pw); void f5(const Widget *pw); catch (Widget w) ... catch (Widget\u0026 w) ... catch (const Widget\u0026 w) ... catch (Widget *pw) ... catch (const Widget *pw) ...   让我们先从相同点谈起。你传递函数参数与异常的途径可以是传值、传递引用或传递指针，这是相同的。但是当你传递参数和异常时，系统所要完成的操作过程则是完全不同的。\n产生这个差异的原因是:你调用函数时，程序的控制权最终还会返回到函数的调用处，但是 当你抛出一个异常时，控制权永远不会回到抛出异常的地方。\n有这样一个函数，参数类型是 Widget，并抛出一个 Widget 类型的异常:\n// 一个函数，从流中读值到 Widget 中\ristream operator(istream\u0026 s, Widget\u0026 w);\rvoid passAndThrowWidget()\r{\rWidget localWidget;\rcin  localWidget; //传递 localWidget 到 operator\rthrow localWidget; // 抛出 localWidget 异常\r}\r 当传递 localWidget 到函数 operator»里，不用进行拷贝操作，而是把 operator»内 的引用类型变量 w 指向 localWidget，任何对 w 的操作实际上都施加到 localWidget 上。\n不论通过传值捕获异常还是通过引用捕获(不能通过 指针捕获这个异常，因为类型不匹配)都将进行 lcalWidget 的拷贝操作，也就说传递到 catch 子句中的是 localWidget 的拷贝。对异常对象进行强制复制拷贝，这个限制有助于我们理解参数传递与抛出异常的第二个差异:抛出异常运行 速度比参数传递要慢。\n必须这么做，因为当 localWidget 离开了生存空间后，其析构函数将被调用。如果把 localWidget 本身(而不是它的拷贝)传递给 catch 子句，这个 子句接收到的只是一个被析构了的 Widget，一个 Widget 的“尸体”。这是无法使用的。因此 C++规范要求被做为异常抛出的对象必须被复制。即使被抛出的对象不会被释放，也会进行拷贝操作。\n例如passAndThrowWidget 函数声明 localWidget 为静态变量(static)，\nvoid passAndThrowWidget()\r{\rstatic Widget localWidget;\rcin  localWidget;\rthrow localWidget;\r}\r 当抛出异常时仍将复制出 localWidget 的一个拷贝。\n即使通过引用来捕获异常，也不能在 catch 块中修改 localWidget;仅仅能修改 localWidget 的拷贝。\n当异常对象被拷贝时，拷贝操作是由对象的拷贝构造函数完成的。该拷贝构造函数是对象的静态类型(static type)所对应类的拷贝构造函数，而不是对象的动态类型(dynamic type)对应类的拷贝构造函数。比如以下这经过少许修改的 passAndThrowWidget:\nclass Widget { ... };\rclass SpecialWidget: public Widget { ... }; void passAndThrowWidget()\r{\rSpecialWidget localSpecialWidget;\r...\rWidget\u0026 rw = localSpecialWidget; // rw 引用 SpecialWidget\rthrow rw; //它抛出一个类型为 Widget // 的异常\r}\r 这里抛出的异常对象是 Widget，即使 rw 引用的是一个 SpecialWidget。因为 rw 的静态类型(statictype)是 Widget，而不是 SpecialWidget。你的编译器根本没有注意到 rw 引用的是一个 SpecialWidget。编译器所注意的是 rw 的静态类型(static type)。这种行为可能与你所期 待的不一样，但是这与在其他情况下 C++中拷贝构造函数的行为是一致的。\n异常是其它对象的拷贝，这个事实影响到你如何在 catch 块中再抛出一个异常。比如下 面这两个 catch 块，乍一看好像一样:\ncatch (Widget\u0026 w)\r{\r...\rthrow; }\rcatch (Widget\u0026 w)\r{\r... throw w;\r}\r 这两个 catch 块的差别在于第一个 catch 块中重新抛出的是当前捕获的异常，而第二个 catch 块中重新抛出的是当前捕获异常的一个新的拷贝。\n第一个块中重新抛出的是当前异常(current exception),无论它是什么类型。 特别是如果这个异常开始就是做为 SpecialWidget 类型抛出的，那么第一个块中传递出去的还是 SpecialWidget 异常，即使 w 的静态类型(static type)是 Widget。这是因为重新抛 出异常时没有进行拷贝操作。\n第二个 catch 块重新抛出的是新异常，类型总是 Widget，因为 w 的静态类型(static type)是 Widget。\n一般来说，你应该用 throw 来重新抛出当前的 异常，因为这样不会改变被传递出去的异常类型，而且更有效率，因为不用生成一个新拷贝。\n让我们测试一下下面这三种用来捕获 Widget 异常的 catch 子句，异常是做为passAndThrowWidgetp 抛出的:\n1 2 3  catch (Widget w) ... catch (Widget\u0026 w) ... catch (const Widget\u0026 w) ...   一个被异常抛出的临时对象可以通过普通的引用捕获;它不需要通过指向 const 对象的引用捕获。在函数调用中不允许转递一个临时对象到一个非 const 引用类型的参数里，但是在异常中却被允许。\n让我们先不管这个差异，回到异常对象拷贝的测试上来。我们知道当用传值的方式传递 函数的参数，我们制造了被传递对象的一个拷贝，并把这个拷贝存储到函数的参数里。同样我们通过传值的方式传递一个异常时，也是这么做的。当我 们这样声明一个 catch 子句时:\ncatch (Widget w) ... // 通过传值捕获  会建立两个被抛出对象的拷贝，一个是所有异常都必须建立的临时对象，第二个是把临 时对象拷贝进 w 中。\n同样，当我们通过引用捕获异常时，\ncatch (Widget\u0026 w) ... // 通过引用捕获\rcatch (const Widget\u0026 w) ... //也通过引用捕获\r 这仍旧会建立一个被抛出对象的拷贝:拷贝同样是一个临时对象。相反当我们通过引用传递函数参数时，没有进行对象拷贝。当抛出一个异常时，系统构造的(以后会析构掉)被抛出对象的拷贝数比以相同对象做为参数传递给函数时构造的拷贝数要多一个。\n我们还没有讨论通过指针抛出异常的情况。不过，通过指针抛出异常与通过指针传递参数是相同的。不论哪种方法都是一个指针的拷贝被传递。\n但你不能认为抛出的指针是一个指向局部对象的指针，因为当异常离开局部变量的生存空间时，该局部变量已经被释放。 Catch 子句将获得一个指向已经不存在的对象的指针。这种行为在设计时应该予以避免。\n对象从函数的调用处传递到函数参数里与从异常抛出点传递到 catch 子句里所采用的 方法不同，这只是参数传递与异常传递的区别的一个方面;第二个差异是在函数调用者或抛 出异常者与被调用者或异常捕获者之间的类型匹配的过程不同。比如在标准数学库(the standard math library)中 sqrt 函数:\ndouble sqrt(double); // from  or   我们能这样计算一个整数的平方根，如下所示:\nint i;\rdouble sqrtOfi = sqrt(i);  毫无疑问，C++允许进行从int到double的隐式类型转换，所以在sqrt的调用中，i被悄悄地转变为 double 类型，并且其返回值也是 double。一般来说，catch 子句匹配异常类型时不会进行这样的转换。见下面的代码:\n1 2 3 4 5 6 7 8 9 10 11  void f(int value) { try { if (someFunction()) { throw value; ... } } catch (double d) { ... } ... }   在 try 块中抛出的 int 异常不会被处理 double 异常的 catch 子句捕获。该子句只能捕 获类型真真正正为 double 的异常，不进行类型转换。因此如果要想捕获 int 异常，必须使 用带有 int 或 int\u0026参数的 catch 子句。\n不过在 catch 子句中进行异常匹配时可以进行两种类型转换。第一种是继承类与基类间 的转换。一个用来捕获基类的 catch 子句也可以处理派生类类型的异常。\n例如在标准 C++库 (STL)定义的异常类层次中的诊断部分\n这种派生类与基类(inheritance_based)间的异常类型转换可以作用于数值、引用以 及指针上:\n1 2 3 4 5 6  catch (runtime_error) ... catch (runtime_error\u0026) ... catch (const runtime_error\u0026) ... catch (runtime_error*) ... catch (const runtime_error*) ...   第二种是允许从一个类型化指针(typed pointer)转变成无类型指针(untyped pointer)，所以带有 const void* 指针的 catch 子句能捕获任何类型的指针类型异常:\ncatch (const void*) ... //捕获任何指针类型异常\r 传递参数和传递异常间最后一点差别是 catch 子句匹配顺序总是取决于它们在程序中出现的顺序。因此一个派生类异常可能被处理其基类异常的 catch 子句捕获，即使同时存在 有能直接处理该派生类异常的 catch 子句，与相同的 try 块相对应。例如:\n1 2 3 4 5 6 7 8 9  try { ... } catch(logic_error\u0026 ex) { ... } catch(invalid_argument\u0026 ex) { ... }   当你调用一个虚拟函数时，被调用的函数位于与发出函数调用的对象的动态类型(dynamic type)最相近的类里。你可以这样说虚拟函数采用最优适合法， 而异常处理采用的是最先适合法。如果一个处理派生类异常的 catch 子句位于处理基类异常 的 catch 子句后面，编译器会发出警告。\n综上所述，把一个对象传递给函数或一个对象调用虚拟函数与把一个对象做为异常抛出，这之间有三个主要区别。\n 异常对象在传递时总被进行拷贝;当通过传值方式捕获 时，异常对象被拷贝了两次。对象做为参数传递给函数时不一定需要被拷贝。 对象做为异常被抛出与做为参数传递给函数相比，前者类型转换比后者要少(前者只有两种转换形式)。 catch 子句进行异常类型匹配的顺序是它们在源代码中出现的顺序，第一个类型匹配成功的 catch 将被用来执行。当一个对象调用一个虚拟函数时，被选择的函数位于与对象类型匹配最佳的类里，即使该类不是在源代码的最前头。  ",
  "wordCount" : "3282",
  "inLanguage": "zh-cn",
  "datePublished": "2017-09-01T17:18:07Z",
  "dateModified": "2017-09-01T17:18:07Z",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "/post/%E6%8A%9B%E5%87%BAexception%E4%B8%8E%E4%BC%A0%E9%80%92%E5%8F%82%E6%95%B0%E6%88%96%E8%B0%83%E7%94%A8%E8%99%9A%E5%87%BD%E6%95%B0%E4%B9%8B%E9%97%B4%E7%9A%84%E5%8C%BA%E5%88%AB/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Forz Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "/favicon.ico"
    }
  }
}
</script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>
<noscript>
    <style type="text/css">
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: #1d1e20;
                --entry: #2e2e33;
                --primary: rgba(255, 255, 255, 0.84);
                --secondary: rgba(255, 255, 255, 0.56);
                --tertiary: rgba(255, 255, 255, 0.16);
                --content: rgba(255, 255, 255, 0.74);
                --hljs-bg: #2e2e33;
                --code-bg: #37383e;
                --border: #333;
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="/" accesskey="h" title="Forz Blog (Alt + H)">Forz Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="/" title="Home">
                    <span>Home</span>
                </a>
            </li>
            <li>
                <a href="/post/" title="Archives">
                    <span>Archives</span>
                </a>
            </li>
            <li>
                <a href="/tags/" title="Tags">
                    <span>Tags</span>
                </a>
            </li>
            <li>
                <a href="/categories/" title="Categories">
                    <span>Categories</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    
    <h1 class="post-title">
      “抛出exception”与“传递参数”或“调用虚函数”之间的区别
    </h1>
    <div class="post-meta">September 1, 2017
</div>
  </header> 
  <div class="post-content"><p>从语法上看，在函数里声明参数与在 catch 子句中声明参数几乎没有什么差别:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-c" data-lang="c"><span class="n">class</span> <span class="n">Widget</span> <span class="p">{</span> <span class="p">...</span> <span class="p">};</span>
<span class="kt">void</span> <span class="nf">f1</span><span class="p">(</span><span class="n">Widget</span> <span class="n">w</span><span class="p">);</span>
<span class="kt">void</span> <span class="nf">f2</span><span class="p">(</span><span class="n">Widget</span><span class="o">&amp;</span> <span class="n">w</span><span class="p">);</span>
<span class="kt">void</span> <span class="nf">f3</span><span class="p">(</span><span class="k">const</span> <span class="n">Widget</span><span class="o">&amp;</span> <span class="n">w</span><span class="p">);</span> 
<span class="kt">void</span> <span class="nf">f4</span><span class="p">(</span><span class="n">Widget</span> <span class="o">*</span><span class="n">pw</span><span class="p">);</span>
<span class="kt">void</span> <span class="nf">f5</span><span class="p">(</span><span class="k">const</span> <span class="n">Widget</span> <span class="o">*</span><span class="n">pw</span><span class="p">);</span> 
<span class="n">catch</span> <span class="p">(</span><span class="n">Widget</span> <span class="n">w</span><span class="p">)</span> <span class="p">...</span>
<span class="n">catch</span> <span class="p">(</span><span class="n">Widget</span><span class="o">&amp;</span> <span class="n">w</span><span class="p">)</span> <span class="p">...</span> 
<span class="n">catch</span> <span class="p">(</span><span class="k">const</span> <span class="n">Widget</span><span class="o">&amp;</span> <span class="n">w</span><span class="p">)</span> <span class="p">...</span> 
<span class="n">catch</span> <span class="p">(</span><span class="n">Widget</span> <span class="o">*</span><span class="n">pw</span><span class="p">)</span> <span class="p">...</span> 
<span class="n">catch</span> <span class="p">(</span><span class="k">const</span> <span class="n">Widget</span> <span class="o">*</span><span class="n">pw</span><span class="p">)</span> <span class="p">...</span>
</code></pre></td></tr></table>
</div>
</div><p>让我们先从相同点谈起。你传递函数参数与异常的途径可以是传值、传递引用或传递指针，这是相同的。但是当你传递参数和异常时，系统所要完成的操作过程则是完全不同的。</p>
<p>产生这个差异的原因是:你调用函数时，程序的控制权最终还会返回到函数的调用处，但是 当你抛出一个异常时，控制权永远不会回到抛出异常的地方。</p>
<p>有这样一个函数，参数类型是 Widget，并抛出一个 Widget 类型的异常:</p>
<pre><code>// 一个函数，从流中读值到 Widget 中
istream operator&gt;&gt;(istream&amp; s, Widget&amp; w);
void passAndThrowWidget()
{
    Widget localWidget;
	cin &gt;&gt; localWidget; //传递 localWidget 到 operator&gt;&gt;
	throw localWidget; // 抛出 localWidget 异常
}
</code></pre>
<p>当传递 localWidget 到函数 operator&raquo;里，不用进行拷贝操作，而是把 operator&raquo;内 的引用类型变量 w 指向 localWidget，任何对 w 的操作实际上都施加到 localWidget 上。</p>
<p>不论通过传值捕获异常还是通过引用捕获(不能通过 指针捕获这个异常，因为类型不匹配)都将进行 lcalWidget 的拷贝操作，也就说传递到 catch 子句中的是 localWidget 的拷贝。对异常对象进行强制复制拷贝，这个限制有助于我们理解参数传递与抛出异常的第二个差异:抛出异常运行 速度比参数传递要慢。</p>
<p>必须这么做，因为当 localWidget 离开了生存空间后，其析构函数将被调用。如果把 localWidget 本身(而不是它的拷贝)传递给 catch 子句，这个 子句接收到的只是一个被析构了的 Widget，一个 Widget 的“尸体”。这是无法使用的。因此 C++规范要求被做为异常抛出的对象必须被复制。即使被抛出的对象不会被释放，也会进行拷贝操作。</p>
<p>例如passAndThrowWidget 函数声明 localWidget 为静态变量(static)，</p>
<pre><code>void passAndThrowWidget()
{
	static Widget localWidget;
	cin &gt;&gt; localWidget;
	throw localWidget;
}
</code></pre>
<p>当抛出异常时仍将复制出 localWidget 的一个拷贝。</p>
<p>即使通过引用来捕获异常，也不能在 catch 块中修改 localWidget;仅仅能修改 localWidget 的拷贝。</p>
<p>当异常对象被拷贝时，拷贝操作是由对象的拷贝构造函数完成的。该拷贝构造函数是对象的静态类型(static type)所对应类的拷贝构造函数，而不是对象的动态类型(dynamic type)对应类的拷贝构造函数。比如以下这经过少许修改的 passAndThrowWidget:</p>
<pre><code>class Widget { ... };
class SpecialWidget: public Widget { ... }; void 	passAndThrowWidget()
{
SpecialWidget localSpecialWidget;
...
Widget&amp; rw = localSpecialWidget; // rw 引用 SpecialWidget
throw rw; //它抛出一个类型为 Widget // 的异常
}
</code></pre>
<p>这里抛出的异常对象是 Widget，即使 rw 引用的是一个 SpecialWidget。因为 rw 的静态类型(statictype)是 Widget，而不是 SpecialWidget。你的编译器根本没有注意到 rw 引用的是一个 SpecialWidget。编译器所注意的是 rw 的静态类型(static type)。这种行为可能与你所期 待的不一样，但是这与在其他情况下 C++中拷贝构造函数的行为是一致的。</p>
<p>异常是其它对象的拷贝，这个事实影响到你如何在 catch 块中再抛出一个异常。比如下 面这两个 catch 块，乍一看好像一样:</p>
<pre><code>catch (Widget&amp; w)
{
...
throw; 
}

catch (Widget&amp; w)
{
... 
throw w;
}
</code></pre>
<p>这两个 catch 块的差别在于第一个 catch 块中重新抛出的是当前捕获的异常，而第二个 catch 块中重新抛出的是当前捕获异常的一个新的拷贝。</p>
<p>第一个块中重新抛出的是当前异常(current exception),无论它是什么类型。 特别是如果这个异常开始就是做为 SpecialWidget 类型抛出的，那么第一个块中传递出去的还是 SpecialWidget 异常，即使 w 的静态类型(static type)是 Widget。这是因为重新抛 出异常时没有进行拷贝操作。</p>
<p>第二个 catch 块重新抛出的是新异常，类型总是 Widget，因为 w 的静态类型(static type)是 Widget。</p>
<p>一般来说，你应该用 throw 来重新抛出当前的 异常，因为这样不会改变被传递出去的异常类型，而且更有效率，因为不用生成一个新拷贝。</p>
<p>让我们测试一下下面这三种用来捕获 Widget 异常的 catch 子句，异常是做为passAndThrowWidgetp 抛出的:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-c" data-lang="c"><span class="n">catch</span> <span class="p">(</span><span class="n">Widget</span> <span class="n">w</span><span class="p">)</span> <span class="p">...</span>
<span class="n">catch</span> <span class="p">(</span><span class="n">Widget</span><span class="o">&amp;</span> <span class="n">w</span><span class="p">)</span> <span class="p">...</span>
<span class="n">catch</span> <span class="p">(</span><span class="k">const</span> <span class="n">Widget</span><span class="o">&amp;</span> <span class="n">w</span><span class="p">)</span> <span class="p">...</span>
</code></pre></td></tr></table>
</div>
</div><p>一个被异常抛出的临时对象可以通过普通的引用捕获;它不需要通过指向 const 对象的引用捕获。在函数调用中不允许转递一个临时对象到一个非 const 引用类型的参数里，但是在异常中却被允许。</p>
<p>让我们先不管这个差异，回到异常对象拷贝的测试上来。我们知道当用传值的方式传递 函数的参数，我们制造了被传递对象的一个拷贝，并把这个拷贝存储到函数的参数里。同样我们通过传值的方式传递一个异常时，也是这么做的。当我 们这样声明一个 catch 子句时:</p>
<pre><code>catch (Widget w) ... // 通过传值捕获 
</code></pre>
<p>会建立<strong>两个</strong>被抛出对象的拷贝，一个是所有异常都必须建立的临时对象，第二个是把临
时对象拷贝进 w 中。</p>
<p>同样，当我们通过引用捕获异常时，</p>
<pre><code>catch (Widget&amp; w) ... // 通过引用捕获
catch (const Widget&amp; w) ... //也通过引用捕获
</code></pre>
<p>这仍旧会建立一个被抛出对象的拷贝:拷贝同样是一个临时对象。相反当我们通过引用传递函数参数时，没有进行对象拷贝。当抛出一个异常时，系统构造的(以后会析构掉)被抛出对象的拷贝数比以相同对象做为参数传递给函数时构造的拷贝数要多一个。</p>
<p>我们还没有讨论通过指针抛出异常的情况。不过，通过指针抛出异常与通过指针传递参数是相同的。不论哪种方法都是一个指针的拷贝被传递。</p>
<p>但你不能认为抛出的指针是一个指向局部对象的指针，因为当异常离开局部变量的生存空间时，该局部变量已经被释放。 Catch 子句将获得一个指向已经不存在的对象的指针。这种行为在设计时应该予以避免。</p>
<p>对象从函数的调用处传递到函数参数里与从异常抛出点传递到 catch 子句里所采用的 方法不同，这只是参数传递与异常传递的区别的一个方面;第二个差异是在函数调用者或抛 出异常者与被调用者或异常捕获者之间的类型匹配的过程不同。比如在标准数学库(the standard math library)中 sqrt 函数:</p>
<pre><code>double sqrt(double); // from &lt;cmath&gt; or &lt;math.h&gt; 
</code></pre>
<p>我们能这样计算一个整数的平方根，如下所示:</p>
<pre><code>int i;
double sqrtOfi = sqrt(i); 
</code></pre>
<p>毫无疑问，C++允许进行从int到double的隐式类型转换，所以在sqrt的调用中，i被悄悄地转变为 double 类型，并且其返回值也是 double。一般来说，catch 子句匹配异常类型时不会进行这样的转换。见下面的代码:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-c" data-lang="c"><span class="kt">void</span> <span class="nf">f</span><span class="p">(</span><span class="kt">int</span> <span class="n">value</span><span class="p">)</span>
<span class="p">{</span>
  <span class="n">try</span> <span class="p">{</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">someFunction</span><span class="p">())</span> <span class="p">{</span>
      <span class="n">throw</span> <span class="n">value</span><span class="p">;</span>
    <span class="p">...</span>
<span class="p">}</span> <span class="p">}</span>
  <span class="n">catch</span> <span class="p">(</span><span class="kt">double</span> <span class="n">d</span><span class="p">)</span> <span class="p">{</span>
    <span class="p">...</span>
<span class="p">}</span>
<span class="p">...</span> <span class="p">}</span>
</code></pre></td></tr></table>
</div>
</div><p>在 try 块中抛出的 int 异常不会被处理 double 异常的 catch 子句捕获。该子句只能捕 获类型真真正正为 double 的异常，不进行类型转换。因此如果要想捕获 int 异常，必须使 用带有 int 或 int&amp;参数的 catch 子句。</p>
<p>不过在 catch 子句中进行异常匹配时可以进行两种类型转换。第一种是继承类与基类间 的转换。一个用来捕获基类的 catch 子句也可以处理派生类类型的异常。</p>
<p>例如在标准 C++库 (STL)定义的异常类层次中的诊断部分</p>
<p><img loading="lazy" src="https://blog-1257321977.cos.ap-beijing.myqcloud.com/QQ20170901-194853@2x.png" alt=""  />
</p>
<p>这种派生类与基类(inheritance_based)间的异常类型转换可以作用于数值、引用以 及指针上:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-c" data-lang="c"><span class="n">catch</span> <span class="p">(</span><span class="n">runtime_error</span><span class="p">)</span> <span class="p">...</span>
<span class="n">catch</span> <span class="p">(</span><span class="n">runtime_error</span><span class="o">&amp;</span><span class="p">)</span> <span class="p">...</span> 
<span class="n">catch</span> <span class="p">(</span><span class="k">const</span> <span class="n">runtime_error</span><span class="o">&amp;</span><span class="p">)</span> <span class="p">...</span>
<span class="n">catch</span> <span class="p">(</span><span class="n">runtime_error</span><span class="o">*</span><span class="p">)</span> <span class="p">...</span> 
<span class="n">catch</span> <span class="p">(</span><span class="k">const</span> <span class="n">runtime_error</span><span class="o">*</span><span class="p">)</span> <span class="p">...</span>

</code></pre></td></tr></table>
</div>
</div><p>第二种是允许从一个类型化指针(typed pointer)转变成无类型指针(untyped
pointer)，所以带有 const void* 指针的 catch 子句能捕获任何类型的指针类型异常:</p>
<pre><code>catch (const void*) ... //捕获任何指针类型异常
</code></pre>
<p>传递参数和传递异常间最后一点差别是 catch 子句匹配顺序总是取决于它们在程序中出现的顺序。因此一个派生类异常可能被处理其基类异常的 catch 子句捕获，即使同时存在 有能直接处理该派生类异常的 catch 子句，与相同的 try 块相对应。例如:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span><span class="lnt">9
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-c" data-lang="c"><span class="n">try</span> <span class="p">{</span>
<span class="p">...</span> 
<span class="p">}</span>
<span class="n">catch</span><span class="p">(</span><span class="n">logic_error</span><span class="o">&amp;</span> <span class="n">ex</span><span class="p">)</span> <span class="p">{</span>
 <span class="p">...</span>
<span class="p">}</span>
<span class="n">catch</span><span class="p">(</span><span class="n">invalid_argument</span><span class="o">&amp;</span> <span class="n">ex</span><span class="p">)</span> <span class="p">{</span>
 <span class="p">...</span>
<span class="p">}</span>
</code></pre></td></tr></table>
</div>
</div><p>当你调用一个虚拟函数时，被调用的函数位于与发出函数调用的对象的动态类型(dynamic type)最相近的类里。你可以这样说虚拟函数采用最优适合法， 而异常处理采用的是最先适合法。如果一个处理派生类异常的 catch 子句位于处理基类异常 的 catch 子句后面，编译器会发出警告。</p>
<p>综上所述，把一个对象传递给函数或一个对象调用虚拟函数与把一个对象做为异常抛出，这之间有三个主要区别。</p>
<ol>
<li>异常对象在传递时总被进行拷贝;当通过传值方式捕获 时，异常对象被拷贝了两次。对象做为参数传递给函数时不一定需要被拷贝。</li>
<li>对象做为异常被抛出与做为参数传递给函数相比，前者类型转换比后者要少(前者只有两种转换形式)。</li>
<li>catch 子句进行异常类型匹配的顺序是它们在源代码中出现的顺序，第一个类型匹配成功的 catch 将被用来执行。当一个对象调用一个虚拟函数时，被选择的函数位于与对象类型匹配最佳的类里，即使该类不是在源代码的最前头。</li>
</ol>


  </div>

  <footer class="post-footer">
  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2021 <a href="/">Forz Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)">
    <button class="top-link" id="top-link" type="button" accesskey="g">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
            <path d="M12 6H0l6-6z" />
        </svg>
    </button>
</a>

<script>
    let menu = document.getElementById('menu')
    menu.scrollLeft = localStorage.getItem("menu-scroll-position");
    menu.onscroll = function () {
        localStorage.setItem("menu-scroll-position", menu.scrollLeft);
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
</body>

</html>
